ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,483 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
+                KeReleaseSpinLock(&PspQuotaLock, OldIrql);
+                return STATUS_QUOTA_EXCEEDED;
+            }
+
+            /* Try to raise the quota limits again */
+            MmRaisePoolQuota(QuotaType,
+                             QuotaBlock->QuotaEntry[QuotaType].Limit,
+                             &UpdatedLimit);
+        }
+
+        /* Release the lock */
         KeReleaseSpinLock(&PspQuotaLock, OldIrql);
-        return STATUS_SUCCESS;
-        //return STATUS_QUOTA_EXCEEDED; /* caller raises the exception */
+
+        /* Enforce a new raised limit */
+        FinalLimit = QuotaBlock->QuotaEntry[QuotaType].Limit - UpdatedLimit;
+        InterlockedCompareExchangeSizeT(&QuotaBlock->QuotaEntry[QuotaType].Limit,
+                                        FinalLimit,
+                                        QuotaBlock->QuotaEntry[QuotaType].Limit);
+
+        /*
+         * Now determine if the current usage and the
+         * amounting by the caller still exceeds the
+         * quota limit of the process. If it's still
+         * over the limit then there's nothing we can
+         * do, so fail.
+         */
+        if (QuotaBlock->QuotaEntry[QuotaType].Usage + Amount >
+            QuotaBlock->QuotaEntry[QuotaType].Limit)
+        {
+            DPRINT1("PspChargeProcessQuotaSpecifiedPool(): Quota amount exceeds the limit (limit -- %lu || amount -- %lu)\n",
+                    QuotaBlock->QuotaEntry[QuotaType].Limit, Amount);
+            return STATUS_QUOTA_EXCEEDED;
+        }
     }
 
-    InterlockedExchangeAdd((LONG*)&Process->QuotaUsage[QuotaType], Amount);
+    /* Update the quota usage */
+    InterlockedCompareExchangeSizeT(&QuotaBlock->QuotaEntry[QuotaType].Usage,
+                                    QuotaBlock->QuotaEntry[QuotaType].Usage + Amount,
+                                    QuotaBlock->QuotaEntry[QuotaType].Usage);
 
-    if (Process->QuotaPeak[QuotaType] < Process->QuotaUsage[QuotaType])
+    /* Are we being given a process as well? */
+    if (Process)
     {
-        Process->QuotaPeak[QuotaType] = Process->QuotaUsage[QuotaType];
+        /* We're being given, check that's not a system one */
+        ASSERT(Process != PsInitialSystemProcess);
+
+        InterlockedExchangeAddSizeT(&Process->QuotaUsage[QuotaType], Amount);
+
+        /*
+         * OK, we've now updated the quota usage of the process
+         * based upon the amount that the caller wanted to charge.
+         * Although the peak of process quota can be less than it was
+         * before so update the peaks as well accordingly.
+         */
+        if (Process->QuotaPeak[QuotaType] < Process->QuotaUsage[QuotaType])
+        {
+            Process->QuotaPeak[QuotaType] = Process->QuotaUsage[QuotaType];
+            Process->QuotaBlock->QuotaEntry[QuotaType].Peak = Process->QuotaBlock->QuotaEntry[QuotaType].Usage;
```

Comment:
Shouldn't this be its own condition? Just because one process hits the peak of its usage doesn't mean the whole quota block is at peak.